sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -0,0 +1,213 @@
+                        .DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model)));
+            }
+
+            // Classic property OR event
+            return propertyReturnType
+                .Union(property.AccessorList.Accessors
+                    .SelectMany(accessor => accessor.DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model))));
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BaseMethodDeclarationSyntax method,
+            SemanticModel model)
+        {
+            return method.DescendantNodes()
+                .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model));
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypesFromVariableDeclarator(VariableDeclaratorSyntax variable,
+            SemanticModel model)
+        {
+            if (variable.Initializer != null)
+            {
+                var typeInfo = model.GetTypeInfo(variable.Initializer.Value);
+                yield return typeInfo.Type;
+                yield return typeInfo.ConvertedType;
+            }
+            else
+            {
+                var variableReturnType = variable.FirstAncestorOrSelf<VariableDeclarationSyntax>()?.Type;
+                if (variableReturnType != null)
+                {
+                    yield return model.GetSymbolInfo(variableReturnType).Symbol as ITypeSymbol;
+                }
+            }
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypeSymbolsFromSyntaxNode(SyntaxNode node, SemanticModel model)
+        {
+            var parameter = node as ParameterSyntax;
+            if (parameter?.Type != null)
+            {
+                yield return model.GetSymbolInfo(parameter.Type).Symbol as ITypeSymbol;
+            }
+
+            var variableDeclaration = node as VariableDeclarationSyntax;
+            if (variableDeclaration != null)
+            {
+                yield return model.GetSymbolInfo(variableDeclaration.Type).Symbol as ITypeSymbol;
+            }
+
+            var objectCreation = node as ObjectCreationExpressionSyntax;
+            if (objectCreation != null)
+            {
+                var objectCreationTypeInfo = model.GetTypeInfo(objectCreation);
+                yield return objectCreationTypeInfo.Type;
+                yield return objectCreationTypeInfo.ConvertedType;
+            }
+
+            var identifierName = node as IdentifierNameSyntax;
+            if (identifierName != null)
+            {
+                yield return model.GetSymbolInfo(identifierName).Symbol as ITypeSymbol;
+            }
+        }
+
+        private static int FilterAndCountCoupling(IEnumerable<ITypeSymbol> types)
+        {
+            return types.Distinct()
+                .WhereNotNull()
+                .Where(type => type.TypeKind != TypeKind.Enum &&
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Drop 'Where' and move the condition into the 'Count'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)